x-pack/plugins/lists/server/services/utils/transform_elastic_named_search_to_list_item.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SearchResponse } from 'elasticsearch';
+import { isEmpty } from 'lodash/fp';
+
+import { SearchEsListItemSchema, SearchListItemArraySchema, Type } from '../../../common/schemas';
+
+import { transformElasticHitsToListItem } from './transform_elastic_to_list_item';
+
+export interface TransformElasticMSearchToListItemOptions {
+  response: SearchResponse<SearchEsListItemSchema>;
+  type: Type;
+  value: unknown[];
+}
+
+/**
+ * Given an ElasticSearch response this will look to see if the named query matches the
+ * index found. The named query will have to be in the format of, "1.0", "1.1", "2.0" where the
+ * major number "1,2,n" will match with the index.
+ * Ref: https://www.elastic.co/guide/en/elasticsearch//reference/7.9/query-dsl-bool-query.html#named-queries
+ * @param response The elastic response
+ * @param type The list type
+ * @param value The values to check against the named queries.
+ */
+export const transformElasticNamedSearchToListItem = ({
+  response,
+  type,
+  value,
+}: TransformElasticMSearchToListItemOptions): SearchListItemArraySchema => {
+  return value.reduce<SearchListItemArraySchema>((accum, singleValue, index) => {
+    const matchingHits = response.hits.hits.filter((hit) => {
+      if (hit.matched_queries != null) {
+        return hit.matched_queries.some((matchedQuery) => {
+          const [matchedQueryIndex] = matchedQuery.split('.');
+          return matchedQueryIndex === `${index}`;
+        });
+      } else {
+        return false;
+      }
+    });
+    if (!isEmpty(matchingHits)) {
+      const items = transformElasticHitsToListItem({ hits: matchingHits, type });
+      const reduced = [
+        ...accum,
+        {
+          items,
+          value: singleValue,
+        },
+      ];
+      return reduced;
+    } else {
+      return accum;
+    }
+  }, []);
```

Comment:
Can we use `value.map` here instead of `value.reduce` and return the (possibly empty) list of matching list items for every value that is passed in? Having a one-to-one correlation between the input values and the objects in the returned array would make it simple to determine if a particular input value is in the list by checking the object in the returned array at the same index. If we remove values that aren't found from the return array then that correlation is lost. Essentially the function would explicitly return whether each input value is in the value list rather than returning only the found values and forcing the caller to determine by omission which values were not found.

I think this would let us remove the additional logic that builds the set of found values and then compares each event to the sets in `createSetToFilterAgainst` and `filterEvents` and instead directly use the return value from this function to filter events. e.g. to determine if `events[n]` needs to be filtered out, we would only have to check the object in `transformElasticNamedSearchToListItem()[n]`. Thoughts?

Reply:
Sounds good! Yeah, I had it initially this way, but then thought maybe that's not what people/devs were looking for, but it sounds like it is, so yeah, I will change this back and see if it clears things up upstream better. Thank you for the hard look and thinking about this and the API's.